From 26a2d88c9694f151da3d4448b5b94f652f21e076 Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Thu, 4 Sep 2025 15:45:00 +0800
Subject: [PATCH 15/32] misc: sli_2700: Fix SLIM calibration rarely fail

Add retry policy for SLIM.

Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
Change-Id: Ie4db5b0992fc28c78ea2d42a78ab57c7c8667aeb
(cherry picked from commit 73ceaefe53163f837af350d478f40d8c666596ea)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/misc/sli_ast2700.c | 83 +++++++++++++++++++++-----------------
 1 file changed, 47 insertions(+), 36 deletions(-)

diff --git a/drivers/misc/sli_ast2700.c b/drivers/misc/sli_ast2700.c
index 33aee03edb..48bdae888d 100644
--- a/drivers/misc/sli_ast2700.c
+++ b/drivers/misc/sli_ast2700.c
@@ -351,7 +351,7 @@ static int sli_calibrate_mbus_pad_delay(struct sli_data *data, int index, int be
 	    data->die0.phy_clk_freq == SLI_PHYCLK_788M)
 		d_def = 5;
 
-	for (count = 0; count < 99; count++) {
+	for (count = 0; count < 50; count++) {
 		for (d = begin; d < end; d++) {
 			sli_set_mbus_delay_single(kx, index, d, is_k_rx);
 
@@ -372,8 +372,9 @@ static int sli_calibrate_mbus_pad_delay(struct sli_data *data, int index, int be
 			}
 		}
 
-		if ((d_last_pass - d_first_pass) > 3)
+		if ((d_last_pass - d_first_pass) >= 3)
 			break;
+		debug("%s SLIM[%d] DS win: {%d, %d} retry %d\n", die_name, index, d_first_pass, d_last_pass, count);
 	}
 
 	if (d_first_pass == -1)
@@ -395,6 +396,7 @@ static void sli_calibrate_mbus_delay(struct sli_data *data, bool is_k_rx)
 	int d_last_pass = -1;
 	int d_def = 12;
 	int win_size = 0;
+	int count;
 	char *die_name = (is_k_rx) ? "IOD" : "CPUD";
 	uintptr_t kx = (is_k_rx) ? data->die1.slim : data->die0.slim;
 
@@ -410,39 +412,46 @@ static void sli_calibrate_mbus_delay(struct sli_data *data, bool is_k_rx)
 		clrbits_le32(kx + SLI_CTRL_I, SLI_RX_PHY_LAH_SEL_NEG);
 
 	/* Find coarse delay */
-	for (dc = SLIM_COARSE_D_BEGIN; dc < SLIM_COARSE_D_END; dc++) {
-		sli_set_mbus_delay(kx, dc, dc, dc, dc, is_k_rx);
+	for (count = 0; count < 50; count++) {
+		for (dc = SLIM_COARSE_D_BEGIN; dc < SLIM_COARSE_D_END; dc++) {
+			sli_set_mbus_delay(kx, dc, dc, dc, dc, is_k_rx);
 
-		/* Reset CPU-die TX and IO-die RX */
-		sli_clear(data->die0.slim, SLI_RESET_TRIGGER);
-		sli_clear(data->die1.slim, SLI_RESET_TRIGGER);
+			/* Reset CPU-die TX and IO-die RX */
+			sli_clear(data->die0.slim, SLI_RESET_TRIGGER);
+			sli_clear(data->die1.slim, SLI_RESET_TRIGGER);
 
-		/* Check result */
-		sli_clear_interrupt_status(data->die1.slim);
-		udelay(200);
-		if (is_sli_suspend(data->die1.slim) > 0) {
-			if (d_first_pass == -1)
-				d_first_pass = dc;
+			/* Check result */
+			sli_clear_interrupt_status(data->die1.slim);
+			udelay(200);
+			if (is_sli_suspend(data->die1.slim) > 0) {
+				if (d_first_pass == -1)
+					d_first_pass = dc;
 
-			d_last_pass = dc;
-		} else if (d_last_pass != -1) {
-			if ((d_last_pass - d_first_pass) > win_size) {
-				win_size = d_last_pass - d_first_pass;
-				sli_log_mbus_pad_delay(data, 0, d_first_pass, d_last_pass);
-				debug("%s SLIM DS coarse win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
+				d_last_pass = dc;
+				debug("%s SLIM DS detect win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
+			} else if (d_last_pass != -1) {
+				if ((d_last_pass - d_first_pass) > win_size) {
+					win_size = d_last_pass - d_first_pass;
+					sli_log_mbus_pad_delay(data, 0, d_first_pass, d_last_pass);
+					debug("%s SLIM DS coarse win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
+				}
+				d_first_pass = -1;
+				d_last_pass = -1;
 			}
-			d_first_pass = -1;
-			d_last_pass = -1;
 		}
-	}
 
-	if (win_size == 0 && d_last_pass != -1) {
-		win_size = d_last_pass - d_first_pass;
-		sli_log_mbus_pad_delay(data, 0, d_first_pass, d_last_pass);
-		debug("%s SLIM DS coarse win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
+		if (win_size == 0 && d_last_pass != -1) {
+			win_size = d_last_pass - d_first_pass;
+			sli_log_mbus_pad_delay(data, 0, d_first_pass, d_last_pass);
+			debug("%s SLIM DS coarse win: {%d, %d}\n", die_name, d_first_pass, d_last_pass);
+		}
+
+		sli_get_mbus_pad_delay(data, 0, &d_first_pass, &d_last_pass);
+		if ((d_last_pass - d_first_pass) >= 3)
+			break;
+		debug("%s SLIM DS retry %d\n", die_name, count);
 	}
 
-	sli_get_mbus_pad_delay(data, 0, &d_first_pass, &d_last_pass);
 	dc = (d_first_pass + d_last_pass) >> 1;
 	if (dc == 0)
 		dc = d_def;
@@ -454,18 +463,20 @@ static void sli_calibrate_mbus_delay(struct sli_data *data, bool is_k_rx)
 	begin = max(dc - SLIM_FINE_MARGIN, 0);
 	end = min(dc + SLIM_FINE_MARGIN, 31);
 
-	/* Fine-tune per-PAD delay */
-	d0 = sli_calibrate_mbus_pad_delay(data, 0, begin, end, is_k_rx);
-	sli_set_mbus_delay_single(kx, 0, d0, is_k_rx);
+	if (win_size) {
+		/* Fine-tune per-PAD delay */
+		d0 = sli_calibrate_mbus_pad_delay(data, 0, begin, end, is_k_rx);
+		sli_set_mbus_delay_single(kx, 0, d0, is_k_rx);
 
-	d1 = sli_calibrate_mbus_pad_delay(data, 1, begin, end, is_k_rx);
-	sli_set_mbus_delay_single(kx, 1, d1, is_k_rx);
+		d1 = sli_calibrate_mbus_pad_delay(data, 1, begin, end, is_k_rx);
+		sli_set_mbus_delay_single(kx, 1, d1, is_k_rx);
 
-	d2 = sli_calibrate_mbus_pad_delay(data, 2, begin, end, is_k_rx);
-	sli_set_mbus_delay_single(kx, 2, d2, is_k_rx);
+		d2 = sli_calibrate_mbus_pad_delay(data, 2, begin, end, is_k_rx);
+		sli_set_mbus_delay_single(kx, 2, d2, is_k_rx);
 
-	d3 = sli_calibrate_mbus_pad_delay(data, 3, begin, end, is_k_rx);
-	sli_set_mbus_delay_single(kx, 3, d3, is_k_rx);
+		d3 = sli_calibrate_mbus_pad_delay(data, 3, begin, end, is_k_rx);
+		sli_set_mbus_delay_single(kx, 3, d3, is_k_rx);
+	}
 
 	/* Reset CPU-die TX and IO-die RX */
 	sli_clear(data->die0.slim, SLI_RESET_TRIGGER);
-- 
2.34.1

